File: gorm/repository.go
package happywakeygorm

import (
	"context"
	"errors"
	"strings"
	"unicode"
	"unicode/utf8"

	"gorm.io/gorm"
)

var ErrInvalidSubject = errors.New("verified subject must be non-empty, bounded, and whitespace-free")

type ReadRepository struct {
	db *gorm.DB
}

func NewReadRepository(db *gorm.DB) (*ReadRepository, error) {
	if db == nil {
		return nil, errors.New("database is required")
	}
	return &ReadRepository{db: db}, nil
}

func (repository *ReadRepository) AlarmsForSubject(ctx context.Context, subject string) ([]Alarm, error) {
	if err := validateSubject(subject); err != nil {
		return nil, err
	}
	var alarms []Alarm
	result := repository.db.WithContext(ctx).
		Where("owner_id = ?", subject).
		Order("created_at ASC").
		Find(&alarms)
	return alarms, result.Error
}

func validateSubject(subject string) error {
	if subject == "" || len(subject) > 512 || !utf8.ValidString(subject) || strings.IndexFunc(subject, func(r rune) bool { return r == '\uFFFD' || unicode.IsSpace(r) }) >= 0 {
		return ErrInvalidSubject
	}
	return nil
}

File: drizzle/src/repository.ts
import { asc, eq } from "drizzle-orm";
import { alarms } from "./schema.js";

const MAX_SUBJECT_BYTES = 512;

/**
 * Return only the predicates and ordering for a subject-scoped alarm read.
 * The web/API adapter owns the concrete database connection and can receive a
 * subject only after Shared Auth has established it.
 */
export function alarmsForSubject(subject: string) {
  validateSubject(subject);
  return {
    where: eq(alarms.ownerId, subject),
    orderBy: asc(alarms.createdAt),
  } as const;
}

export function validateSubject(subject: string): void {
  if (
    subject.length === 0 ||
    Buffer.byteLength(subject, "utf8") > MAX_SUBJECT_BYTES ||
    /\s/u.test(subject)
  ) {
    throw new TypeError(
      "verified subject must be non-empty, bounded, and whitespace-free",
    );
  }
}
